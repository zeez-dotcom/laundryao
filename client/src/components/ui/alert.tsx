import * as React from "react"
import { cva, type VariantProps } from "class-variance-authority"

import { cn } from "@/lib/utils"

const alertVariants = cva(
  "relative w-full rounded-lg border p-4 [&>svg~*]:pl-7 [&>svg+div]:translate-y-[-3px] [&>svg]:absolute [&>svg]:left-4 [&>svg]:top-4 [&>svg]:text-foreground",
  {
    variants: {
      variant: {
        default: "bg-background text-foreground",
        destructive:
          "border-destructive/50 text-destructive dark:border-destructive [&>svg]:text-destructive",
      },
    },
    defaultVariants: {
      variant: "default",
    },
  }
)

const Alert = React.forwardRef<
  HTMLDivElement,
  React.HTMLAttributes<HTMLDivElement> & VariantProps<typeof alertVariants>
>(({ className, variant, ...props }, ref) => (
  <div
    ref={ref}
    role="alert"
    className={cn(alertVariants({ variant }), className)}
    {...props}
  />
))
Alert.displayName = "Alert"

const AlertTitle = React.forwardRef<
  HTMLParagraphElement,
  React.HTMLAttributes<HTMLHeadingElement>
>(({ className, ...props }, ref) => (
  <h5
    ref={ref}
    className={cn("mb-1 font-medium leading-none tracking-tight", className)}
    {...props}
  />
))
AlertTitle.displayName = "AlertTitle"

const AlertDescription = React.forwardRef<
  HTMLDivElement,
  React.HTMLAttributes<HTMLDivElement> & { copyable?: boolean }
>(({ className, children, copyable, ...props }, ref) => {
  const localRef = React.useRef<HTMLDivElement | null>(null)
  const contentRef = (node: HTMLDivElement) => {
    localRef.current = node
    if (typeof ref === 'function') ref(node as any)
    else if (ref) (ref as any).current = node
  }
  const isCopyable = copyable || (props as any)["data-copyable"] !== undefined
  const onCopy = async () => {
    const text = localRef.current?.innerText || ""
    try { await navigator.clipboard.writeText(text) } catch {}
  }
  return (
    <div
      ref={contentRef}
      className={cn("text-sm [&_p]:leading-relaxed select-text relative pr-10", className)}
      {...props}
    >
      {children}
      {isCopyable && (
        <button
          type="button"
          onClick={onCopy}
          className="absolute top-0 right-0 text-xs text-muted-foreground hover:text-foreground underline"
          aria-label="Copy error"
        >
          Copy
        </button>
      )}
    </div>
  )
})
AlertDescription.displayName = "AlertDescription"

export { Alert, AlertTitle, AlertDescription }
