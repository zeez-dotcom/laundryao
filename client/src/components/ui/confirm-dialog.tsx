import {
  AlertDialog,
  AlertDialogContent,
  AlertDialogHeader,
  AlertDialogTitle,
  AlertDialogDescription,
  AlertDialogFooter,
  AlertDialogCancel,
  AlertDialogAction,
} from "@/components/ui/alert-dialog";

interface ConfirmDialogProps {
  open: boolean;
  onOpenChange: (open: boolean) => void;
  title: string;
  description?: string;
  confirmText: string;
  cancelText: string;
  onConfirm: () => void;
}

export function ConfirmDialog({
  open,
  onOpenChange,
  title,
  description,
  confirmText,
  cancelText,
  onConfirm,
}: ConfirmDialogProps) {
  return (
    <AlertDialog open={open} onOpenChange={onOpenChange}>
      <AlertDialogContent>
        <AlertDialogHeader>
          <AlertDialogTitle>{title}</AlertDialogTitle>
          {description && (
            <AlertDialogDescription>{description}</AlertDialogDescription>
          )}
        </AlertDialogHeader>
        <AlertDialogFooter>
          <AlertDialogCancel>{cancelText}</AlertDialogCancel>
          <AlertDialogAction onClick={onConfirm}>{confirmText}</AlertDialogAction>
        </AlertDialogFooter>
      </AlertDialogContent>
    </AlertDialog>
  );
}

export default ConfirmDialog;
