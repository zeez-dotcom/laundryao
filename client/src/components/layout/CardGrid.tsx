import { ReactNode, useCallback, useEffect, useMemo, useState } from "react";
import { Card, CardContent, CardDescription, CardFooter, CardHeader, CardTitle } from "@/components/ui/card";
import { Accordion, AccordionContent, AccordionItem, AccordionTrigger } from "@/components/ui/accordion";
import { Checkbox } from "@/components/ui/checkbox";
import { Badge } from "@/components/ui/badge";
import { cn } from "@/lib/utils";

export type CardChecklistItem = {
  id: string;
  label: string;
  description?: string;
  initiallyChecked?: boolean;
};

export type CardAccordionSection = {
  id: string;
  title: string;
  summary?: string;
  content: ReactNode;
  defaultOpen?: boolean;
};

export type CardGridCard = {
  id: string;
  title: string;
  description?: string;
  badgeLabel?: string;
  icon?: ReactNode;
  accent?: "primary" | "secondary" | "neutral";
  actions?: ReactNode;
  footer?: ReactNode;
  checklist?: CardChecklistItem[];
  persistChecklistKey?: string;
  accordionSections?: CardAccordionSection[];
};

export type CardGridProps = {
  cards: CardGridCard[];
  className?: string;
  columns?: {
    base?: number;
    md?: number;
    lg?: number;
  };
};

const STORAGE_PREFIX = "laundryao.card-grid";

type ChecklistState = Record<string, boolean>;

function usePersistentChecklist(key: string | undefined, items: CardChecklistItem[] | undefined) {
  const storageKey = key ? `${STORAGE_PREFIX}.${key}` : undefined;
  const [state, setState] = useState<ChecklistState>(() => {
    if (!storageKey || typeof window === "undefined") {
      return {};
    }
    try {
      const raw = localStorage.getItem(storageKey);
      if (raw) {
        return JSON.parse(raw) as ChecklistState;
      }
    } catch (error) {
      console.warn("Failed to read checklist state", error);
    }
    if (!items) return {};
    return Object.fromEntries(items.map((item) => [item.id, Boolean(item.initiallyChecked)]));
  });

  useEffect(() => {
    if (!storageKey || typeof window === "undefined") return;
    try {
      localStorage.setItem(storageKey, JSON.stringify(state));
    } catch (error) {
      console.warn("Failed to persist checklist state", error);
    }
  }, [state, storageKey]);

  // Ensure new checklist items are available when definitions change
  useEffect(() => {
    if (!items) return;
    setState((current) => {
      const next = { ...current };
      let changed = false;
      for (const item of items) {
        if (!(item.id in next)) {
          next[item.id] = Boolean(item.initiallyChecked);
          changed = true;
        }
      }
      return changed ? next : current;
    });
  }, [items]);

  const toggle = useCallback((itemId: string, checked: boolean) => {
    setState((prev) => ({ ...prev, [itemId]: checked }));
  }, []);

  return { state, toggle };
}

const GRID_CLASS_MAP: Record<number, string> = {
  1: "grid-cols-1",
  2: "grid-cols-2",
  3: "grid-cols-3",
  4: "grid-cols-4",
};

function gridClass(count: number, prefix?: string) {
  const safeCount = Math.max(1, Math.min(4, Math.floor(count)));
  const baseClass = GRID_CLASS_MAP[safeCount] ?? GRID_CLASS_MAP[1];
  return prefix ? `${prefix}:${baseClass}` : baseClass;
}

function buildGridClass(columns: CardGridProps["columns"] | undefined) {
  const base = columns?.base ?? 1;
  const md = columns?.md ?? (base > 1 ? base : 2);
  const lg = columns?.lg ?? Math.max(md, 3);
  return cn("grid w-full", gridClass(base), md > 1 ? gridClass(md, "md") : undefined, lg > 1 ? gridClass(lg, "xl") : undefined);
}

function CardGridItem({ card }: { card: CardGridCard }) {
  const { state, toggle } = usePersistentChecklist(card.persistChecklistKey ?? card.id, card.checklist);

  return (
    <Card
      id={`card-${card.id}`}
      className={cn(
        "flex h-full flex-col shadow-sm transition-shadow focus-within:ring-2 focus-within:ring-[var(--focus)]",
        card.accent === "primary" && "border-primary/40 shadow-[var(--shadow-soft)]",
        card.accent === "secondary" && "border-secondary/40",
      )}
    >
      <CardHeader className="space-y-3">
        <div className="flex items-start justify-between gap-3">
          <div className="flex items-center gap-3">
            {card.icon ? (
              <div className="flex size-10 items-center justify-center rounded-full bg-primary/10 text-primary">{card.icon}</div>
            ) : null}
            <div>
              <CardTitle className="font-heading text-[var(--text-lg)] leading-[var(--line-height-tight)]">
                {card.title}
              </CardTitle>
              {card.description ? (
                <CardDescription className="text-[var(--text-sm)] leading-[var(--line-height-relaxed)]">
                  {card.description}
                </CardDescription>
              ) : null}
            </div>
          </div>
          {card.badgeLabel ? (
            <Badge variant={card.accent === "secondary" ? "secondary" : "default"}>{card.badgeLabel}</Badge>
          ) : null}
        </div>
        {card.actions ? <div className="flex flex-wrap gap-2">{card.actions}</div> : null}
      </CardHeader>

      {card.accordionSections?.length ? (
        <CardContent className="space-y-2">
          <Accordion
            type="multiple"
            defaultValue={card.accordionSections.filter((section) => section.defaultOpen).map((section) => section.id)}
            className="w-full"
          >
            {card.accordionSections.map((section) => (
              <AccordionItem key={section.id} value={section.id}>
                <AccordionTrigger className="text-left text-[var(--text-md)]">
                  <div className="flex flex-col items-start gap-1">
                    <span>{section.title}</span>
                    {section.summary ? (
                      <span className="text-[var(--text-sm)] text-muted-foreground">{section.summary}</span>
                    ) : null}
                  </div>
                </AccordionTrigger>
                <AccordionContent className="pt-2">{section.content}</AccordionContent>
              </AccordionItem>
            ))}
          </Accordion>
        </CardContent>
      ) : null}

      {card.checklist?.length ? (
        <CardContent className="mt-auto space-y-3 rounded-lg bg-[var(--surface-muted)] p-4">
          <div className="text-[var(--text-sm)] font-medium text-foreground">Contextual checklist</div>
          <ul className="space-y-2">
            {card.checklist.map((item) => (
              <li key={item.id} className="flex items-start gap-3 rounded-md bg-background/60 p-3">
                <Checkbox
                  id={`${card.id}-${item.id}`}
                  checked={Boolean(state[item.id])}
                  onCheckedChange={(checked) => toggle(item.id, Boolean(checked))}
                  aria-describedby={item.description ? `${card.id}-${item.id}-description` : undefined}
                  className="mt-1"
                />
                <div>
                  <label htmlFor={`${card.id}-${item.id}`} className="text-[var(--text-sm)] font-medium text-foreground">
                    {item.label}
                  </label>
                  {item.description ? (
                    <p id={`${card.id}-${item.id}-description`} className="text-[var(--text-xs)] text-muted-foreground">
                      {item.description}
                    </p>
                  ) : null}
                </div>
              </li>
            ))}
          </ul>
        </CardContent>
      ) : null}

      {card.footer ? <CardFooter className="mt-auto flex flex-wrap gap-2">{card.footer}</CardFooter> : null}
    </Card>
  );
}

export function CardGrid({ cards, className, columns }: CardGridProps) {
  const gridClassName = useMemo(() => buildGridClass(columns), [columns]);

  return (
    <div
      className={cn(gridClassName, "gap-[var(--space-lg)]", className)}
      style={{
        gap: "var(--space-lg, 1.5rem)",
      }}
      data-testid="card-grid"
    >
      {cards.map((card) => (
        <CardGridItem key={card.id} card={card} />
      ))}
    </div>
  );
}

export default CardGrid;
