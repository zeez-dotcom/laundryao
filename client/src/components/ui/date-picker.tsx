import * as React from "react";
import { Popover, PopoverContent, PopoverTrigger } from "@/components/ui/popover";
import { Button } from "@/components/ui/button";
import { Calendar } from "@/components/ui/calendar";
import { Calendar as CalendarIcon } from "lucide-react";
import { format } from "date-fns";

interface DatePickerProps {
  date?: Date;
  onChange?: (date?: Date) => void;
  placeholder?: string;
}

export function DatePicker({ date, onChange, placeholder = "Pick a date" }: DatePickerProps) {
  const [open, setOpen] = React.useState(false);

  return (
    <Popover open={open} onOpenChange={setOpen}>
      <PopoverTrigger asChild>
        <Button variant="outline" className="justify-start w-full">
          <CalendarIcon className="mr-2 h-4 w-4" />
          {date ? format(date, "MMM dd, yyyy") : <span className="text-muted-foreground">{placeholder}</span>}
        </Button>
      </PopoverTrigger>
      <PopoverContent className="p-0" align="start">
        <Calendar
          mode="single"
          selected={date}
          onSelect={(d) => {
            onChange?.(d ?? undefined);
            setOpen(false);
          }}
        />
      </PopoverContent>
    </Popover>
  );
}

