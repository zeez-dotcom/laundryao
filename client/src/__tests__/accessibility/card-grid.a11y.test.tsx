import { render } from "@testing-library/react";
import { axe } from "jest-axe";
import { describe, it, expect } from "vitest";
import CardGrid, { type CardGridCard } from "@/components/layout/CardGrid";
import { Button } from "@/components/ui/button";

describe("CardGrid accessibility", () => {
  const cards: CardGridCard[] = [
    {
      id: "demo",
      title: "Demo card",
      description: "A sample card used for accessibility validation.",
      icon: <span aria-hidden="true">üîç</span>,
      accordionSections: [
        {
          id: "summary",
          title: "Summary",
          summary: "High level summary",
          defaultOpen: true,
          content: <p>Progressive disclosure keeps this content tucked away until needed.</p>,
        },
        {
          id: "actions",
          title: "Actions",
          summary: "Available tasks",
          content: <Button variant="outline">Run task</Button>,
        },
      ],
    },
  ];

  it("passes axe checks", async () => {
    const { container } = render(<CardGrid cards={cards} columns={{ base: 1 }} />);
    expect(await axe(container)).toHaveNoViolations();
    expect(container).toMatchSnapshot();
  });
});
