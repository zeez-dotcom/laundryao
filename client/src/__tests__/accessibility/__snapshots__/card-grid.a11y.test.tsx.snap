// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardGrid accessibility > passes axe checks 1`] = `
<div>
  <div
    class="grid w-full grid-cols-1 md:grid-cols-2 xl:grid-cols-3 gap-[var(--space-lg)]"
    data-testid="card-grid"
    style="gap: var(--space-lg, 1.5rem);"
  >
    <div
      class="rounded-lg border bg-card text-card-foreground flex h-full flex-col shadow-sm transition-shadow focus-within:ring-2 focus-within:ring-[var(--focus)]"
      id="card-demo"
    >
      <div
        class="flex flex-col p-6 space-y-3"
      >
        <div
          class="flex items-start justify-between gap-3"
        >
          <div
            class="flex items-center gap-3"
          >
            <div
              class="flex size-10 items-center justify-center rounded-full bg-primary/10 text-primary"
            >
              <span
                aria-hidden="true"
              >
                üîç
              </span>
            </div>
            <div>
              <div
                class="text-2xl font-semibold tracking-tight font-heading text-[var(--text-lg)] leading-[var(--line-height-tight)]"
              >
                Demo card
              </div>
              <div
                class="text-sm text-[var(--text-sm)] leading-[var(--line-height-relaxed)]"
              >
                A sample card used for accessibility validation.
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="p-6 pt-0 space-y-2"
      >
        <div
          class="w-full"
          data-orientation="vertical"
        >
          <div
            class="border-b"
            data-orientation="vertical"
            data-state="open"
          >
            <h3
              class="flex"
              data-orientation="vertical"
              data-state="open"
            >
              <button
                aria-controls="radix-:r1:"
                aria-expanded="true"
                class="flex flex-1 items-center justify-between py-4 font-medium transition-all hover:underline [&[data-state=open]>svg]:rotate-180 text-left text-[var(--text-md)]"
                data-orientation="vertical"
                data-radix-collection-item=""
                data-state="open"
                id="radix-:r0:"
                type="button"
              >
                <div
                  class="flex flex-col items-start gap-1"
                >
                  <span>
                    Summary
                  </span>
                  <span
                    class="text-[var(--text-sm)] text-muted-foreground"
                  >
                    High level summary
                  </span>
                </div>
                <svg
                  class="lucide lucide-chevron-down h-4 w-4 shrink-0 transition-transform duration-200"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m6 9 6 6 6-6"
                  />
                </svg>
              </button>
            </h3>
            <div
              aria-labelledby="radix-:r0:"
              class="overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
              data-orientation="vertical"
              data-state="open"
              id="radix-:r1:"
              role="region"
              style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width); transition-duration: 0s; animation-name: none;"
            >
              <div
                class="pb-4 pt-2"
              >
                <p>
                  Progressive disclosure keeps this content tucked away until needed.
                </p>
              </div>
            </div>
          </div>
          <div
            class="border-b"
            data-orientation="vertical"
            data-state="closed"
          >
            <h3
              class="flex"
              data-orientation="vertical"
              data-state="closed"
            >
              <button
                aria-controls="radix-:r3:"
                aria-expanded="false"
                class="flex flex-1 items-center justify-between py-4 font-medium transition-all hover:underline [&[data-state=open]>svg]:rotate-180 text-left text-[var(--text-md)]"
                data-orientation="vertical"
                data-radix-collection-item=""
                data-state="closed"
                id="radix-:r2:"
                type="button"
              >
                <div
                  class="flex flex-col items-start gap-1"
                >
                  <span>
                    Actions
                  </span>
                  <span
                    class="text-[var(--text-sm)] text-muted-foreground"
                  >
                    Available tasks
                  </span>
                </div>
                <svg
                  class="lucide lucide-chevron-down h-4 w-4 shrink-0 transition-transform duration-200"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m6 9 6 6 6-6"
                  />
                </svg>
              </button>
            </h3>
            <div
              aria-labelledby="radix-:r2:"
              class="overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
              data-orientation="vertical"
              data-state="closed"
              hidden=""
              id="radix-:r3:"
              role="region"
              style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
